```
drivers/ext4grpquota/ext4_grpquota.go:179:1:warning: exported method Ext4GrpQuota.GetNextQuotaID should have comment or be unexported (golint)
```

--- a/.gitignore
+++ b/.gitignore
@@ -4,7 +4,7 @@
 *.dll
 *.so
 *.dylib
-./bin
+bin/
 
 # Test binary, build with `go test -c`
 *.test

--- a/cmd/setquota/main.go
+++ b/cmd/setquota/main.go
@@ -117,8 +117,8 @@ func setquota(context *cli.Context) error {
 			"set one of 'bhard' or 'bsoft' or 'ihard' or 'isoft'")
 	}
 	limit := &types.QuotaLimit{
-		BlockHardLimit: bhard,
-		BlockSoftLimit: bsoft,
+		BlockHardLimit: bhard * 1024 * 1024,
+		BlockSoftLimit: bsoft * 1024 * 1024,
 		InodeHardLimit: ihard,
 		InodeSoftLimit: isoft,
 	}
@@ -126,6 +126,13 @@ func setquota(context *cli.Context) error {
 	quota, err := libquota.NewQuota(file)
 	if err != nil {
 		return err
+	}
+
+	if id == 0 {
+		id, err = quota.GetQuotaID(file)
+		if err != nil {
+			return err
+		}
 	}
 
 	return quota.SetQuota(file, id, limit)

--- a/drivers/ext4grpquota/ext4_grpquota.go
+++ b/drivers/ext4grpquota/ext4_grpquota.go
@@ -1,6 +1,21 @@
 package ext4grpquota
 
-import "github.com/pouchcontainer/libquota/types"
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/pouchcontainer/libquota/pkg/cmd"
+	"github.com/pouchcontainer/libquota/pkg/fs"
+	"github.com/pouchcontainer/libquota/types"
+
+	"github.com/gofrs/flock"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
 
 // Ext4GrpQuota defines the ext4 group quota struct
 type Ext4GrpQuota struct {
@@ -10,12 +25,112 @@ type Ext4GrpQuota struct {
 // New is used to check whether support to use ext4 group quota,
 // and returns the ext4 group quota object.
 func New(file string) (*Ext4GrpQuota, error) {
-	// TODO: validate fs info
-	return &Ext4GrpQuota{}, nil
+	// get mountpoint
+	mount, err := fs.GetMountPoint(file)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get file(%s) mountpoint", file)
+	}
+
+	if !strings.Contains(mount.Opts, "grpquota") {
+		res, err := cmd.Run(0, "mount",
+			"-o", "remount,grpquota",
+			mount.Device, mount.MountPoint)
+		if err != nil {
+			return nil, err
+		}
+
+		if res.ExitCode != 0 {
+			return nil, errors.Errorf("failed to remount with prjquota, result(%v)", res)
+		}
+	}
+
+	//
+	vfsVersion, quotaFilename, err := getVFSVersionAndQuotaFile(mount.MountPoint)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get vfs version and quota file")
+	}
+
+	filename := mount.MountPoint + "/" + quotaFilename
+	if _, err := os.Stat(filename); err != nil {
+		os.Remove(mount.MountPoint + "/aquota.user")
+
+		header := []byte{0x27, 0x19, 0xc0, 0xd9, 0x00, 0x00, 0x00, 0x00, 0x80, 0x3a, 0x09, 0x00, 0x80,
+			0x3a, 0x09, 0x00, 0x00, 0x00, 0x00, 0x00, 0x06, 0x00, 0x00, 0x00, 0x00, 0x00,
+			0x00, 0x00, 0x05, 0x00, 0x00, 0x00}
+		if vfsVersion == "vfsv1" {
+			header[4] = 0x01
+		}
+
+		if writeErr := ioutil.WriteFile(filename, header, 0644); writeErr != nil {
+			return nil, errors.Wrapf(writeErr, "failed to write file, filename: (%s), vfs version: (%s)",
+				filename, vfsVersion)
+		}
+		if res, err := cmd.Run(0, "setquota", "-g", "-t", "43200", "43200", mount.MountPoint); err != nil {
+			os.Remove(filename)
+			return nil, errors.Wrapf(err, "failed to setquota, result(%v)", res)
+		}
+		if res, err := cmd.Run(0, "setquota", "-g", "0", "0", "0", "0", "0", mount.MountPoint); err != nil {
+			os.Remove(filename)
+			return nil, errors.Wrapf(err, "failed to setquota, result(%v)", res)
+		}
+	}
+
+	// check group quota status, on or not, pay attention, the right exit code of command 'quotaon' is '1'.
+	res, err := cmd.Run(0, "quotaon", "-pg", mount.MountPoint)
+	if err != nil && res.ExitCode != 1 {
+		return nil, errors.Wrapf(err, "failed to quota on for mountpoint(%s), result(%v)",
+			mount.MountPoint, res)
+	}
+
+	// set quota on
+	if !strings.Contains(res.Stdout, " is on") {
+		if res, err = cmd.Run(0, "quotaon", mount.MountPoint); err != nil {
+			return nil, errors.Wrapf(err, "failed to quotaon, mountpoint(%s), result(%v)",
+				mount.MountPoint, res)
+		}
+	}
+
+	idFLock := flock.New(filepath.Join(mount.MountPoint, "quota-id.lock"))
+	idFLock.Lock()
+	defer idFLock.Unlock()
+
+	idNext, idMap, err := loadQuotaIDs(mount.MountPoint, "-gn")
+	if err != nil {
+		return nil, err
+	}
+
+	return &Ext4GrpQuota{
+		BaseQuota: types.BaseQuota{
+			Mount:   mount,
+			IDFLock: idFLock,
+			IDNext:  idNext,
+			IDMap:   idMap,
+		},
+	}, nil
 }
 
 // SetQuota is used to set the file's ext4 group quota with quota id.
 func (q *Ext4GrpQuota) SetQuota(file string, id uint64, quota *types.QuotaLimit) error {
+	if file == "" || id == 0 || quota == nil {
+		return errors.Errorf("invalid arguments, file(%s), id(%d), quota(%v)", file, id, quota)
+	}
+
+	// set quota id on files
+	args := fmt.Sprintf("-n system.subtree -v %d %s", id, file)
+	res, err := cmd.Run(0, "setfattr", strings.Fields(args)...)
+	if err != nil || res.ExitCode != 0 {
+		return errors.Wrapf(err, "failed to set quota, mountpoint(%s), result(%v)", q.Mount.MountPoint, res)
+	}
+
+	// set quota on mountpoint
+	args = fmt.Sprintf("-g %d %d %d %d %d %s",
+		id, quota.BlockSoftLimit/1024, quota.BlockHardLimit/1024, quota.InodeSoftLimit, quota.InodeHardLimit, q.Mount.MountPoint)
+
+	res, err = cmd.Run(0, "setquota", strings.Fields(args)...)
+	if err != nil || res.ExitCode != 0 {
+		return errors.Wrapf(err, "failed to set quota, mountpoint(%s), result(%v)", q.Mount.MountPoint, res)
+	}
+
 	return nil
 }
 
@@ -26,5 +141,153 @@ func (q *Ext4GrpQuota) GetQuota(file string) (*types.QuotaLimit, error) {
 
 // GetQuotaID returns the file's ext4 group quota id.
 func (q *Ext4GrpQuota) GetQuotaID(file string) (uint64, error) {
-	return 0, nil
+	var (
+		getNextID bool
+		id        uint64
+	)
+
+	if file == "" {
+		getNextID = true
+	}
+
+	if _, err := os.Stat(file); err != nil {
+		return 0, err
+	}
+
+	res, err := cmd.Run(0, "getfattr", "-n", "system.subtree", "--only-values", "--absolute-names", file)
+	if err == nil && res.ExitCode == 0 {
+		num, err := strconv.Atoi(res.Stdout)
+		if err != nil {
+			return 0, err
+		}
+		id = uint64(num)
+
+		if id == 0 {
+			getNextID = true
+		}
+	} else {
+		getNextID = true
+	}
+
+	if getNextID {
+		id = q.GetNextQuotaID()
+	}
+
+	return id, nil
+}
+
+// GetNextQuotaID returns the quota of this mountpoint that can be used.
+func (q *Ext4GrpQuota) GetNextQuotaID() uint64 {
+	q.IDFLock.Lock()
+	defer func() {
+		q.IDNext++
+		q.IDFLock.Unlock()
+	}()
+
+	return q.IDNext
+}
+
+func getVFSVersionAndQuotaFile(file string) (string, string, error) {
+	devID, err := fs.GetDevID(file)
+	if err != nil {
+		return "", "", err
+	}
+
+	output, err := ioutil.ReadFile(fs.ProcMount)
+	if err != nil {
+		logrus.Warnf("failed to read file: (%s), err: (%v)", fs.ProcMount, err)
+		return "", "", errors.Wrap(err, "failed to read /proc/mounts")
+	}
+
+	vfsVersion := "vfsv0"
+	quotaFilename := "aquota.group"
+	for _, line := range strings.Split(string(output), "\n") {
+		// TODO: add an example here to make following code readable.
+		// /dev/sdb1 /home/pouch ext4 rw,relatime,prjquota,data=ordered 0 0 ?
+		parts := strings.Split(line, " ")
+		if len(parts) != 6 {
+			continue
+		}
+
+		devID2, _ := fs.GetDevID(parts[1])
+		if devID != devID2 {
+			continue
+		}
+
+		for _, opt := range strings.Split(parts[3], ",") {
+			items := strings.SplitN(opt, "=", 2)
+			if len(items) != 2 {
+				continue
+			}
+			switch items[0] {
+			case "jqfmt":
+				vfsVersion = items[1]
+			case "grpjquota":
+				quotaFilename = items[1]
+			}
+		}
+		return vfsVersion, quotaFilename, nil
+	}
+
+	return vfsVersion, quotaFilename, nil
+}
+
+func loadQuotaIDs(mountpoint string, opts string) (uint64, map[uint64]*types.QuotaLimit, error) {
+	idMap := make(map[uint64]*types.QuotaLimit)
+
+	res, err := cmd.Run(0, "repquota", opts, mountpoint)
+	if err != nil || res.ExitCode != 0 {
+		return 0, nil, errors.Wrapf(err, "failed to execute [repquota %s %s], result(%v)",
+			opts, mountpoint, res)
+	}
+
+	// example output:
+	//                          Block limits                File limits
+	// Group           used    soft    hard  grace    used  soft  hard  grace
+	// ----------------------------------------------------------------------
+	// #0        -- 12693804       0       0         163764     0     0
+	// #16777217 --      44       0 10485760             18     0     0
+
+	currentID := types.BaseQuotaID
+	lines := strings.Split(res.Stdout, "\n")
+	for _, line := range lines {
+		// find all lines with prefix '#'
+		if len(line) == 0 || line[0] != '#' {
+			continue
+		}
+
+		parts := strings.Fields(line)
+		// right line is:
+		// #16777217 --      44       0 10485760             18     0     0
+		if len(parts) != 8 {
+			continue
+		}
+
+		id, err := strconv.Atoi(parts[0][1:])
+		if err != nil {
+			continue
+		}
+
+		qid := uint64(id)
+		if qid < types.BaseQuotaID {
+			continue
+		}
+
+		if qid > currentID {
+			currentID = qid
+		}
+
+		bsLimit, _ := strconv.Atoi(parts[3])
+		bhLimit, _ := strconv.Atoi(parts[4])
+		isLimit, _ := strconv.Atoi(parts[6])
+		ihLimit, _ := strconv.Atoi(parts[7])
+		idMap[qid] = &types.QuotaLimit{
+			BlockSoftLimit: uint64(bsLimit) * 1024,
+			BlockHardLimit: uint64(bhLimit) * 1024,
+			InodeSoftLimit: uint64(isLimit),
+			InodeHardLimit: uint64(ihLimit),
+		}
+	}
+
+	return currentID + 1, idMap, nil
 }

--- a/drivers/ext4prjquota/ext4_prjquota.go
+++ b/drivers/ext4prjquota/ext4_prjquota.go
@@ -1,6 +1,14 @@
 package ext4prjquota
 
-import "github.com/pouchcontainer/libquota/types"
+import (
+	"strings"
+
+	"github.com/pouchcontainer/libquota/pkg/cmd"
+	"github.com/pouchcontainer/libquota/pkg/fs"
+	"github.com/pouchcontainer/libquota/types"
+
+	"github.com/pkg/errors"
+)
 
 // Ext4PrjQuota defines the ext4 project quota struct
 type Ext4PrjQuota struct {
@@ -10,8 +18,40 @@ type Ext4PrjQuota struct {
 // New is used to check whether support to use ext4 project quota,
 // and returns the ext4 project quota object.
 func New(file string) (*Ext4PrjQuota, error) {
-	// TODO: validate fs info
-	return &Ext4PrjQuota{}, nil
+	// get mountpoint
+	mount, err := fs.GetMountPoint(file)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get file(%s) mountpoint", file)
+	}
+
+	if !strings.Contains(mount.Opts, "prjquota") {
+		res, err := cmd.Run(0, "mount",
+			"-o", "remount,prjquota",
+			mount.Device, mount.MountPoint)
+		if err != nil {
+			return nil, err
+		}
+
+		if res.ExitCode != 0 {
+			return nil, errors.Errorf("failed to remount with prjquota, result(%v)", res)
+		}
+	}
+
+	// use tool quotaon to set disk quota for mountpoint
+	res, err := cmd.Run(0, "quotaon", "-P", mount.MountPoint)
+	if err != nil {
+		if !strings.Contains(res.Stderr, " File exists") {
+			return nil, err
+		}
+	}
+
+	return &Ext4PrjQuota{
+		BaseQuota: types.BaseQuota{
+			Mount:  mount,
+			IDNext: types.BaseQuotaID,
+			IDMap:  make(map[uint64]*types.QuotaLimit),
+		},
+	}, nil
 }
 
 // SetQuota is used to set the file's ext4 project quota with quota id.

--- a/drivers/xfsprjquota/xfs_prjquota.go
+++ b/drivers/xfsprjquota/xfs_prjquota.go
@@ -54,19 +54,15 @@ func New(file string) (*XfsPrjQuota, error) {
 
 	return &XfsPrjQuota{
 		BaseQuota: types.BaseQuota{
-			IDMap: make(map[uint64]*types.QuotaLimit),
+			Mount:  mount,
+			IDNext: types.BaseQuotaID,
+			IDMap:  make(map[uint64]*types.QuotaLimit),
 		},
 	}, nil
 }
 
 // SetQuota is used to set the file's xfs project quota with quota id.
 func (q *XfsPrjQuota) SetQuota(file string, id uint64, quota *types.QuotaLimit) error {
-	// get mountpoint
-	mount, err := fs.GetMountPoint(file)
-	if err != nil {
-		return errors.Wrapf(err, "failed to get file(%s) mountpoint", file)
-	}
-
 	// set quota id
 	setID := fmt.Sprintf("project -s -p %s %d", file, id)
 
@@ -83,7 +79,7 @@ func (q *XfsPrjQuota) SetQuota(file string, id uint64, quota *types.QuotaLimit) 
 	setLimit := fmt.Sprintf("limit -p bhard=%dm bsoft=%dm ihard=%d isoft=%d %d %s",
 		quota.BlockHardLimit, quota.BlockSoftLimit,
 		quota.InodeHardLimit, quota.InodeSoftLimit,
-		id, mount.MountPoint)
+		id, q.Mount.MountPoint)
 	ret, err = cmd.Run(0, XfsQuota, "-xc", setLimit)
 	if err != nil {
 		return errors.Wrapf(err, "failed to run (%s -xc %s)", XfsQuota, setLimit)
@@ -104,7 +100,5 @@ func (q *XfsPrjQuota) GetQuota(file string) (*types.QuotaLimit, error) {
 
 // GetQuotaID returns the file's xfs project quota id.
 func (q *XfsPrjQuota) GetQuotaID(file string) (uint64, error) {
-	// TODO: Not implemented
-	// xfsctl(path, fd, XFS_IOC_FSGETXATTR, &fsx)
 	return 0, nil
 }

--- a/pkg/cmd/command.go
+++ b/pkg/cmd/command.go
@@ -1,12 +1,19 @@
 package cmd
 
 import (
+	"bufio"
+	"bytes"
 	"fmt"
-	"io/ioutil"
 	"os/exec"
+	"strings"
 	"syscall"
 	"time"
+
+	"github.com/sirupsen/logrus"
 )
+
+// DefaultTimeOut represents the default timeout of running command.
+const DefaultTimeOut = 3600
 
 // CommandResult defines the execute command's result
 type CommandResult struct {
@@ -18,9 +25,15 @@ type CommandResult struct {
 // Run is used to execute command with timeout
 func Run(timeout int, bin string, args ...string) (*CommandResult, error) {
 	var (
-		err      error
-		exitCode int
+		err       error
+		exitCode  int
+		stdoutBuf bytes.Buffer
+		stderrBuf bytes.Buffer
 	)
+
+	if timeout <= 0 {
+		timeout = DefaultTimeOut
+	}
 	cmd := exec.Command(bin, args...)
 
 	stdoutPipe, err := cmd.StdoutPipe()
@@ -41,6 +54,12 @@ func Run(timeout int, bin string, args ...string) (*CommandResult, error) {
 
 	done := make(chan error)
 	go func() {
+		if _, err := bufio.NewReader(stdoutPipe).WriteTo(&stdoutBuf); err != nil {
+			done <- err
+		}
+		if _, err := bufio.NewReader(stderrPipe).WriteTo(&stderrBuf); err != nil {
+			done <- err
+		}
 		done <- cmd.Wait()
 	}()
 
@@ -56,18 +75,12 @@ func Run(timeout int, bin string, args ...string) (*CommandResult, error) {
 		}
 	}
 
-	stdout, err := ioutil.ReadAll(stdoutPipe)
-	if err != nil {
-		return nil, err
-	}
-
-	stderr, err := ioutil.ReadAll(stderrPipe)
-	if err != nil {
-		return nil, err
-	}
+	logrus.Debugf("success to run [ %s %s ], result [%s %s]",
+		bin, strings.Join(args, " "), stdoutBuf.String(), stderrBuf.String())
 
 	return &CommandResult{
-		Stdout:   string(stdout),
-		Stderr:   string(stderr),
-		ExitCode: exitCode}, nil
+		Stdout:   stdoutBuf.String(),
+		Stderr:   stderrBuf.String(),
+		ExitCode: exitCode,
+	}, nil
 }

--- a/pkg/fs/device.go
+++ b/pkg/fs/device.go
@@ -1,0 +1,16 @@
+package fs
+
+import (
+	"syscall"
+
+	"github.com/pkg/errors"
+)
+
+// GetDevID returns device id via syscall according to the input directory.
+func GetDevID(dir string) (uint64, error) {
+	var st syscall.Stat_t
+	if err := syscall.Stat(dir, &st); err != nil {
+		return 0, errors.Wrapf(err, "failed to get device id of directory: (%s)", dir)
+	}
+	return st.Dev, nil
+}

--- a/pkg/fs/mountpoint.go
+++ b/pkg/fs/mountpoint.go
@@ -2,6 +2,7 @@ package fs
 
 import (
 	"io/ioutil"
+	"os"
 	"strings"
 
 	"github.com/pouchcontainer/libquota/pkg/cmd"
@@ -44,7 +45,7 @@ func GetMountPoint(file string) (*Mount, error) {
 	// Filesystem     1K-blocks  Used Available Use% Mounted on
 	// /dev/sdb        41922560 32940  41889620   1% /mnt/data
 	lines := strings.Split(res.Stdout, "\n")
-	if len(lines) != 2 {
+	if len(lines) != 3 {
 		return nil, errors.Errorf("failed to use df to get mountpoint, "+
 			"invalid output(%s)", res.Stdout)
 	}
@@ -66,13 +67,21 @@ func GetMountPoint(file string) (*Mount, error) {
 		if len(parts) != 6 {
 			continue
 		}
+
+		// check device is exist or not.
+		if _, err := os.Stat(parts[0]); err != nil {
+			continue
+		}
+
 		if parts[1] == mountpoint {
-			mount.Device = parts[0]
-			mount.MountPoint = parts[1]
-			mount.FSType = parts[2]
-			mount.Opts = parts[3]
-			mount.Dump = parts[4]
-			mount.Pass = parts[5]
+			mount = &Mount{
+				Device:     parts[0],
+				MountPoint: parts[1],
+				FSType:     parts[2],
+				Opts:       parts[3],
+				Dump:       parts[4],
+				Pass:       parts[5],
+			}
 			break
 		}
 	}

--- a/quota.go
+++ b/quota.go
@@ -17,7 +17,7 @@ func NewQuota(file string) (Quota, error) {
 
 	mount, err := fs.GetMountPoint(file)
 	if err != nil {
-		return nil, errors.Errorf("failed to get (%s) file system information", file)
+		return nil, errors.Wrapf(err, "failed to get file(%s) system information", file)
 	}
 
 	quota = QuotaMapGet(mount.MountPoint)

--- a/types/type.go
+++ b/types/type.go
@@ -1,6 +1,9 @@
 package types
 
-import "sync"
+import (
+	"github.com/gofrs/flock"
+	"github.com/pouchcontainer/libquota/pkg/fs"
+)
 
 // FSType defines the type of file system struct
 type FSType string
@@ -11,18 +14,25 @@ const (
 
 	// Xfs defines the xfs file system
 	Xfs FSType = "xfs"
+
+	// BaseQuotaID represents the minimize quota id.
+	// The value is unit32(2^24).
+	BaseQuotaID = uint64(16777216)
 )
 
 // BaseQuota defines the basic attribute of quota
 type BaseQuota struct {
-	IDMap     map[uint64]*QuotaLimit
-	IDMapLock sync.Mutex
+	Mount   *fs.Mount
+	IDFLock *flock.Flock
+
+	IDNext uint64
+	IDMap  map[uint64]*QuotaLimit
 }
 
 // QuotaLimit defines the attribute of quota limit
 type QuotaLimit struct {
-	BlockSoftLimit uint64
-	BlockHardLimit uint64
+	BlockSoftLimit uint64 // unit is bytes
+	BlockHardLimit uint64 // unit is bytes
 
 	InodeSoftLimit uint64
 	InodeHardLimit uint64

